org/mozilla/jss/util/NativeEnclosure.java

```diff
@@ -0,0 +1,46 @@
+package org.mozilla.jss.util;
+
+/**
+ * Classes extending NativeEnclsoure wrap a single NativeProxy instance,
+ * allowing it to be accessed from the JNI layer but be allocated and scoped
+ * from the Java layer.
+ *
+ * This allows us to data parameters in Java and call open(...) when all
+ * necessary variables have been created. At the end of this object's use,
+ * close must be called to free allocated memory.
+ *
+ * Because this class implements AutoCloseable, it is suggested to add
+ * constructors to derived classes which call open; this'll allow a single
+ * try-with-resources block to scope the lifetime of this object:
+ *
+ * <pre>
+ * try (NEC obj = new NEC(...)) {
+ *      // ... do something with obj ...
+ * }
+ * </pre>
+ */
+public abstract class NativeEnclosure implements AutoCloseable {
+    public NativeProxy ptr;
+    public long ptr_size;
+
+    public final void open() throws Exception {
+        if (ptr == null) {
+            acquireNativeResources();
+        }
+    }
+
+    @Deprecated
+    protected void finalize() throws Throwable {
+        close();
+    }
+
+    public final void close() throws Exception {
+        if (ptr != null) {
+            releaseNativeResources();
+            ptr.close();
```

Comment:
Similarly, who is responsible to nullify `ptr` after `ptr.close()`, or does it matter? Do we need to add this?
```
ptr = null;
```

Reply:
You're right here, I forgot to `null` it.